guides/source/upgrading_ruby_on_rails.md

```diff
@@ -169,6 +169,86 @@ more meaningful preview images. Previously the first frame of the video would be
+times in apps that rely on active storage to serve their images. 
+
+The `:mini_magick` option is not being deprecated, so it is fine to keep using it.
+
+Migrating to libvips requires changing existing image transformation code to the
+`image_processing` macros, and replacing ImageMagick's options with libvips' options.
+
+#### Replace resize with resize_to_limit
+```ruby
+variant(resize: "100x")
+variant(resize_to_limit: [100, nil])
+```
+
+If you forget to do this, when you switch to vips you will see this error: `no implicit conversion to float from string`.
+
+#### Use an array when cropping
+```
+variant(crop: "1920x1080+0+0")
+variant(crop: [0, 0, 1920, 1080])
+```
+
+If you forget to do this, when you switch to vips you will see this error: `unable to call crop: you supplied 2 arguments, but operation needs 5`.
+
+#### Clamp your crop values:
+Vips is more strict than ImageMagick when it comes to cropping:
+1. It will not crop if `x` and/or `y` are negative values. e.g.: `[-10, -10, 100, 100]`
+2. It will not crop if position (`x` or `y`) plus crop dimension (`width`, `height`) is larger than the image. e.g.: a 125x125 image and a crop of `[50, 50, 100, 100]`
+
+If you forget to do this, when you switch to vips you will see this error: `extract_area: bad extract area`
+
+#### Adjust the background color used in resize and pad
+Vips uses black as the default background color `resize_and_pad`, instead of white like ImageMagick. Fix that by using the `background` option:
+```ruby
+variant(resize_and_pad: [300, 300])
+variant(resize_and_pad: [300, 300, background: [255]])
+```
+
+#### Remove any EXIF based rotation
+Vips will auto rotate images using the EXIF value when processing variants. If you were storing rotation values from user uploaded photos to apply rotation with ImageMagick, you must stop doing that:
+```ruby
+variant(format: :jpg, rotate: rotation_value)
+variant(format: :jpg)
+```
+
+#### Replace monochrome with colourspace
+Vips uses a different option to make monochrome images:
+```ruby
+variant(monochrome: true)
+variant(colourspace: "b-w")
+```
+
+#### Switch to libvips options
+Vips uses different options to compress images:
+
+```ruby
+# JPEG
+variant(strip: true, quality: 80, interlace: "JPEG", sampling_factor: "4:2:0", colorspace: "sRGB")
+variant(saver: { strip: true, quality: 80, interlace: true })
+
+# PNG
+variant(strip: true, quality: 75)
+variant(saver: { strip: true, compression: 9 })
+
+# WEBP
+variant(strip: true, quality: 75, define: { webp: { lossless: false, alpha_quality: 85, thread_level: 1 } })
+variant(saver: { strip: true, quality: 75, lossless: false, alpha_q: 85, reduction_effort: 6, smart_subsample: true })
+
+# GIF
+variant(layers: "Optimize")
+variant(saver: { optimize_gif_frames: true, optimize_gif_transparency: true })
+```
+
+#### Clear your caches
+After you deploy the new to production, clear your fragment caches. Active Storage encodes into the url of the image 
```

Comment:
> clear your fragment caches

how?

Reply:
(link to relevant docs would be ideal)

Reply:
I think the easiest would be to advise them to `Rails.cache.clear`. I've checked the caching guide, but didn't find any reference on how to clear the caches. Did you have any other specific doc in mind?

Reply:
oh, that's what you meant by fragment caches. I don't think `Rails.cache.clear` is good upgrade advice.

I believe the cache keys are usually generated based on a hash of the view file's contents. Would that fix things here?

Reply:
Unfortunately not. The call to `.variant` might be inside a helper or a model, so the contents of the partial will not change. I agree that `Rails.cache.clear` is overkill but it's pretty much the only reliable way I've found to ensure all fragments are removed. We might be able to improve this by iterating on all cache entries and only removing those whose keys indicate they are HTML fragments, but I'm unsure of whats the best way to do this.

Reply:
Actually, "clear your fragment caches" is wrong. An app could have a piece of code like this:
```
Rails.cache.fetch("#{id}/photos") do
  photos.map{|photo| photo.file.variant([resize_to_limit: 100])}
end
```

Either devs do a full cache clear or they hunt down by hand each and every instance of `variant` in their app and clear those specific caches.

Reply:
Right, but we can't really tell people to clear their entire Rails cache. It could be enormous and could hold critical data.

I think the better thing to do here is explain a bit more what's going on and suggest improvements. For example, the `cache` view helper method takes a list of args; you could pass a version number `cache("v2")`. That's not a very good solution, I'm open to better ones. Or alternatively we just state the problem and users work it out based on their codebase.

Reply:
Ok, I hadn’t considered that apps might use caches to hold critical data because to me cache = discardable. 

Frankly, I don’t use much caching so I’m out of depth here (as you noticed from my “solution” of doing a full cache clear). 

At this point best I can do is the second choice: state the problem, probably in a little more detail, but let users decide how to fix.